### ISSUE TITLE
IPV6 redirection with xrootd Doors

### ISSUE THREAD

**Issue by cfgamboa** [dCache/dcache#6602]
Dear all,

I am trying to understand why from the LXPLUS node at CERN I am getting the following results:

I am using XROOTD door on dCache to copy a file from. The door is on dcqos002.usatlas.bnl.gov
Are hosts on the same subnet. It should interact with external clients ( like this one) with the external interface.
dcqos001.usatlas.bnl.gov
dcqos002.usatlas.bnl.gov

1. First test ( the file is stored in a pool in the same server as the Xrootd door), dcqos002.usatlas.bnl.gov

```
Pool file location
[dcqos001] (Xrootd-dcqos002@xrootd-dcqos002Domain) admin > \sn cacheinfoof /pnfs/usatlas.bnl.gov/diskdata/test2.dat.fts.2
 dcqos002_24

Copy
[cgamboa@lxplus777 ~]$ xrdcp -f root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.fts.2 /tmp/samepoolhost

[1.23GB/1.23GB][100%][==================================================][30.71MB/s] 
```

2.Second test (the file is stored in a pool in the a different  server as the Xrootd door), dcqos001.usatlas.bnl.gov

```
[dcqos001] (Xrootd-dcqos002@xrootd-dcqos002Domain) admin > \sn cacheinfoof pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp 
 dcqos001_11

Copy
[cgamboa@lxplus777 ~]$ xrdcp -f root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp /tmp/differentpoolhost
[0B/0B][100%][==================================================][0B/s]  
Run: [ERROR] Invalid redirect URL:  (source)
```
Looking at the debug messages for the transfer using `-d 3` this can be highlighted, here the redirection used the IPV6
`**[2022-04-20 23:32:50.203674 +0200][Error  ][XRootD            ] [dcqos002.usatlas.bnl.gov:1096] Got invalid redirection URL: 2620:0:210:1:0:0:0:b?org.dcache.xrootd.client=cgamboa.11944@lxplus777.cern.ch&org.dcache.uuid=5d28f4a3-805d-45cf-b3eoticed**`

Here the door shows this:
```
20 Apr 2022 18:08:55 [xrootd-net-35] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Time since last server cert refresh 171462
20 Apr 2022 18:08:55 [xrootd-net-35] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Loading server certificates. Current refresh interval: 60000 ms
20 Apr 2022 18:08:55 [xrootd-net-35] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Verifying host certificate
20 Apr 2022 18:08:56 [xrootd-net-35] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
20 Apr 2022 18:08:56 [xrootd-net-35] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Time since last server cert refresh 1115
20 Apr 2022 18:08:56 [xrootd-net-35] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Using padded DH secret generation.
20 Apr 2022 18:08:56 [pool-7-thread-42] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp for read
20 Apr 2022 18:08:56 [pool-7-thread-42] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] OPAQUE : {}
20 Apr 2022 18:08:56 [pool-7-thread-42] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Unable to resolve IP address 2620:0:210:1:0:0:0:b to a canonical name
20 Apr 2022 18:08:56 [pool-7-thread-42] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Redirecting to 2620:0:210:1:0:0:0:b, 2620:0:210:1:0:0:0:b/2620:0:210:1:0:0:0:b:33128
20 Apr 2022 18:08:56 [pool-7-thread-42] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHT1Ay8A] Sending the following host information to the client: 2620:0:210:1:0:0:0:b
```

3. Third test, I repeat test 2 but instead I force the transfer to use IPV4, this WORKS
``` [cgamboa@lxplus777 ~]$ xrdcp  -f -DSNetworkStack IPv4 root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp /tmp/differentpoolhost_ipv4
[17B/17B][100%][==================================================][5B/s]  
```

Here the redirection uses the IPV4 address 
```
[2022-04-21 00:06:05.920938 +0200][Debug  ][File              ] [0x1958be0@root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp?xrdcl.requuid=cdcb6d14-e057-4c49-89e5-4c71c3304a10] Sending a read command for handle 0x0 to 192.12.15.11:33194
```

the door shows this:
```
20 Apr 2022 18:11:01 [xrootd-net-36] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Time since last server cert refresh 126458
20 Apr 2022 18:11:01 [xrootd-net-36] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Loading server certificates. Current refresh interval: 60000 ms
20 Apr 2022 18:11:01 [xrootd-net-36] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Verifying host certificate
20 Apr 2022 18:11:03 [xrootd-net-36] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
20 Apr 2022 18:11:03 [xrootd-net-36] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Time since last server cert refresh 1847
20 Apr 2022 18:11:03 [xrootd-net-36] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Using padded DH secret generation.
20 Apr 2022 18:11:03 [pool-7-thread-44] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp for read
20 Apr 2022 18:11:03 [pool-7-thread-44] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] OPAQUE : {}
20 Apr 2022 18:11:03 [pool-7-thread-44] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Unable to resolve IP address 192.12.15.11 to a canonical name
20 Apr 2022 18:11:03 [pool-7-thread-44] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Redirecting to 192.12.15.11, 192.12.15.11/192.12.15.11:33128
20 Apr 2022 18:11:03 [pool-7-thread-44] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA] Sending the following host information to the client: 192.12.15.11
20 Apr 2022 18:11:04 [Xrootd-dcqos002-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdHUTKmwA dcqos001_11 DoorTransferFinished 0000DBC6E3CF2C53497ABDF155B4C59A62B5] Transfer 0000DBC6E3CF2C53497ABDF155B4C59A62B5@PoolName=dcqos001_11 PoolAddress=dcqos001_11@dcqos001elevenDomain finished
```

Could you please advise? It is not clear to me why the transfer fails if the two hosts can resolve the IPV4 and IPV6 addresses at the OS level.

All the best,
Carlos

**Comment by alrossi** [comment]
Carlos,  could you maybe attach the door logs for the first test (the successful one)?

Thanks.

**Comment by cfgamboa** [comment]
Here it is:
```
21 Apr 2022 15:49:23 [xrootd-net-27] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Time since last server cert refresh 75905
21 Apr 2022 15:49:23 [xrootd-net-27] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Loading server certificates. Current refresh interval: 60000 ms
21 Apr 2022 15:49:23 [xrootd-net-27] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Verifying host certificate
21 Apr 2022 15:49:25 [xrootd-net-27] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
21 Apr 2022 15:49:25 [xrootd-net-27] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Time since last server cert refresh 1521
21 Apr 2022 15:49:25 [xrootd-net-27] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Using padded DH secret generation.
21 Apr 2022 15:49:25 [pool-7-thread-53] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.fts.2 for read
21 Apr 2022 15:49:25 [pool-7-thread-53] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] OPAQUE : {}
21 Apr 2022 15:49:25 [pool-7-thread-53] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Redirecting to dcqos002.usatlas.bnl.gov, dcqos002.usatlas.bnl.gov/2620:0:210:1:0:0:0:f:33314
21 Apr 2022 15:49:25 [pool-7-thread-53] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA] Sending the following host information to the client: dcqos002.usatlas.bnl.gov
21 Apr 2022 15:50:14 [Xrootd-dcqos002-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdL2ghCHA dcqos002_24 DoorTransferFinished 00005ACC02C1991641C69D57E8E0C979205B] Transfer 00005ACC02C1991641C69D57E8E0C979205B@PoolName=dcqos002_24 PoolAddress=dcqos002_24@dcqos002twentyfourDomain finished

```

**Comment by alrossi** [comment]
Thank you.  As you can see, in the first case, the host name and not the IP address is being returned.  This is correct behavior as far as the xrootd client is concerned.

A bit of history (I'm including the info below more for our reference than anything else):

Back two years ago we changed the door to send all redirect messages using the host name, because TLS requires this.

```
            /*  According to https://tools.ietf.org/html/rfc5280#section-4.2.1.6
             *  an IP is required to be in the list of Subject Alternative Names
             *  in the host certificate, but these are rarely added in practice.
             *  TLS enforces the RFC and this is a workaround.
             */
```

I had a conversation with Andy about this in which he confirmed that the hostname, not IPs should be used:

```
Indeed, we use the curl algorithm as it is a strict interpretation of
the RFC. ...
a) it's commonly accepted that all certs should have a proper SAN, and
b) redirecting to an IP address is considered bad form as putting IP
addresses (v4 and v6 to boot) in the SAN is practically unmanageable.

Additionally, using DNS blows away all the security since it allows MIN
attacks. BTW we do allow using DNS for GSI authentication to be backward
compatible but TLS is brand new and, technically, there are no backward
compatibility issues here.

So, unfortunately, the RFC says that, yes, you must have the IP address in
the SAN if you redirect using an IP address. But the question is why
don't you use the DNS name? All xroot clients happily accept DNS names in
the redirect and this would make the verification work, as expected. The
xroot protocol won't be the only one you will have to deal with as https
will give you the same headache.

```

I am unsure exactly what is happening in the three scenarios to resolve the redirect endpoint differently, but the xrootd door makes a best effort to return the resolved name, and gives a warning if it does not resolve properly.

```
            String host = address.getHostName();
            if (InetAddresses.isInetAddress(host)) {
                _log.warn("Unable to resolve IP address {} "
                      + "to a canonical name", host);
            }
```

I don't currently have an answer, but could this difference depend on the way your network is configured?

Al

**Comment by alrossi** [comment]
(P.S., while you are not using TLS, the lookup procedure is uniform in the door, so it happens with GSI as well)

**Comment by cfgamboa** [comment]
Hello Al,

I think I was able to identify this issue. This Xrootd fails to initiate a redirection using IPV6 IP address.
At CERN the host/pools hosting the data has the following DNS resolution:
```
[cgamboa@lxplus730 ~]$ nslookup dcqos001.usatlas.bnl.gov
Server:		127.0.0.1
Address:	127.0.0.1#53

Non-authoritative answer:
Name:	dcqos001.usatlas.bnl.gov
Address: 192.12.15.11
Name:	dcqos001.usatlas.bnl.gov
Address: 2620:0:210:1::b
```

I have enabled systematically in the `/etc/hosts` file the resolution for dcqos01.usatlas.bnl.gov the two IPs shown above (IPv4 and IPv6). Tests can be summarize as follows: 

- Test 4. **Succeeded** Here there is not specific resolution for IPv4, it is using the hosts native resolution

```[cgamboa@lxplus730 ~]$ xrdcp -f -DSNetworkStack IPv4 root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp /tmp/differentpoolhost
[17B/17B][100%][==================================================][8B/s]  
```
Note that from the Door log the value of 
`22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Unable to resolve IP address 192.12.15.11 to a canonical name
22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Redirecting to 192.12.15.11, 192.12.15.11/192.12.15.11:33282
`


```
22 Apr 2022 14:41:59 [xrootd-net-2] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Time since last server cert refresh 1230658
22 Apr 2022 14:41:59 [xrootd-net-2] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Loading server certificates. Current refresh interval: 60000 ms
22 Apr 2022 14:41:59 [xrootd-net-2] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Verifying host certificate
22 Apr 2022 14:42:00 [xrootd-net-2] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
22 Apr 2022 14:42:00 [xrootd-net-2] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Time since last server cert refresh 1037
22 Apr 2022 14:42:00 [xrootd-net-2] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Using padded DH secret generation.
22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp for read
22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] OPAQUE : {}
22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Unable to resolve IP address 192.12.15.11 to a canonical name
22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Redirecting to 192.12.15.11, 192.12.15.11/192.12.15.11:33282
22 Apr 2022 14:42:00 [pool-7-thread-3] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg] Sending the following host information to the client: 192.12.15.11
22 Apr 2022 14:42:01 [Xrootd-dcqos002-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQpTnWXg dcqos001_11 DoorTransferFinished 0000DBC6E3CF2C53497ABDF155B4C59A62B5] Transfer 0000DBC6E3CF2C53497ABDF155B4C59A62B5@PoolName=dcqos001_11 PoolAddress=dcqos001_11@dcqos001elevenDomain finished
```


- Test 5. **Succeeded** Enable the resolution of the dcqos001.usatlas.bnl.gov on IPV4 
 I restarted the door on dcqos002. It seems that there is caching and the new values are refreshed when rebooting, I guess  I'm not patience for the `networkaddress.cache.ttl` to kick and load new changes.
The IP is enabled at `/etc/hosts` as `192.12.15.11                    dcqos001.usatlas.bnl.gov        dcqos001`

```
[cgamboa@lxplus730 ~]$ xrdcp -f -DSNetworkStack IPv4 root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp /tmp/differentpoolhost
[17B/17B][100%][==================================================][8B/s]  

Note that on this case the client is able to get the canonical name
`22 Apr 2022 14:47:02 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Redirecting to dcqos001.usatlas.bnl.gov, dcqos001.usatlas.bnl.gov/192.12.15.11:33117
`

```
22 Apr 2022 14:47:00 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Time since last server cert refresh 1650653220623
22 Apr 2022 14:47:00 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Loading server certificates. Current refresh interval: 60000 ms
22 Apr 2022 14:47:00 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Verifying host certificate
22 Apr 2022 14:47:01 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
22 Apr 2022 14:47:01 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Time since last server cert refresh 755
22 Apr 2022 14:47:01 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Using padded DH secret generation.
22 Apr 2022 14:47:02 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp for read
22 Apr 2022 14:47:02 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] OPAQUE : {}
22 Apr 2022 14:47:02 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Redirecting to dcqos001.usatlas.bnl.gov, dcqos001.usatlas.bnl.gov/192.12.15.11:33117
22 Apr 2022 14:47:02 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg] Sending the following host information to the client: dcqos001.usatlas.bnl.gov
22 Apr 2022 14:47:02 [Xrootd-dcqos002-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQqbZCdg dcqos001_11 DoorTransferFinished 0000DBC6E3CF2C53497ABDF155B4C59A62B5] Transfer 0000DBC6E3CF2C53497ABDF155B4C59A62B5@PoolName=dcqos001_11 PoolAddress=dcqos001_11@dcqos001elevenDomain finished
```

Test 6. **FAILS** It seems that the default IPprotocol  at cern LXPLUS is IPV6. If I do not enable the IPV6 on `/etc/hosts` then the transfers fails 
```
[cgamboa@lxplus730 ~]$ xrdcp -f -DSNetworkStack IPv6 root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp /tmp/differentpoolhost
[0B/0B][100%][==================================================][0B/s]  
Run: [ERROR] Invalid redirect URL:  (source)
```
At the door, note that in the case of IPV4 the transfer works even without been able to get the canonical name. Here this is not the case, see the extract
`22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Unable to resolve IP address 2620:0:210:1:0:0:0:b to a canonical name
22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Redirecting to 2620:0:210:1:0:0:0:b, 2620:0:210:1:0:0:0:b/2620:0:210:1:0:0:0:b:33153
`
Door log
```
22 Apr 2022 14:50:59 [xrootd-net-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Time since last server cert refresh 238572
22 Apr 2022 14:50:59 [xrootd-net-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Loading server certificates. Current refresh interval: 60000 ms
22 Apr 2022 14:50:59 [xrootd-net-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Verifying host certificate
22 Apr 2022 14:50:59 [xrootd-net-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
22 Apr 2022 14:50:59 [xrootd-net-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Time since last server cert refresh 630
22 Apr 2022 14:51:00 [xrootd-net-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Using padded DH secret generation.
22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp for read
22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] OPAQUE : {}
22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Unable to resolve IP address 2620:0:210:1:0:0:0:b to a canonical name
22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Redirecting to 2620:0:210:1:0:0:0:b, 2620:0:210:1:0:0:0:b/2620:0:210:1:0:0:0:b:33153
22 Apr 2022 14:51:00 [pool-7-thread-4] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQrUVKGA] Sending the following host information to the client: 2620:0:210:1:0:0:0:b
```


Test 7. **SUCCEEDED**! Enabling the IPV6 IP on the `/etc/hosts` help the xrootd client to create the transfer.
`2620:0:210:1::b                 dcqos001.usatlas.bnl.gov        dcqos001` value in the `/etc/hosts`

```
[cgamboa@lxplus730 ~]$ xrdcp -f -DSNetworkStack IPv6 root://dcqos002.usatlas.bnl.gov:1096/pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp /tmp/differentpoolhost
[17B/17B][100%][==================================================][5B/s]  
```

Door
```
22 Apr 2022 14:55:10 [pool-6-thread-1] [NIC auto-discovery] Adding [fe80::9a03:9bff:fe04:736, dcqos002.usatlas.bnl.gov/10.42.38.49, dcqos002.usatlas.bnl.gov/[2620:0:210:1::f], fe80::9a03:9bff:fe89:e1fe, dcqos002.usatlas.bnl.gov/192.12.15.15]
22 Apr 2022 14:55:36 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Time since last server cert refresh 1650653736862
22 Apr 2022 14:55:36 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Loading server certificates. Current refresh interval: 60000 ms
22 Apr 2022 14:55:36 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Verifying host certificate
22 Apr 2022 14:55:37 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Client protocol version was 10600, using GSIPost49ServerRequestHandler.
22 Apr 2022 14:55:37 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Time since last server cert refresh 673
22 Apr 2022 14:55:37 [xrootd-net-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Using padded DH secret generation.
22 Apr 2022 14:55:38 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Opening pnfs/usatlas.bnl.gov/diskdata/test2.dat.srmcp for read
22 Apr 2022 14:55:38 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] OPAQUE : {}
22 Apr 2022 14:55:38 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Redirecting to dcqos001.usatlas.bnl.gov, dcqos001.usatlas.bnl.gov/2620:0:210:1:0:0:0:b:33153
22 Apr 2022 14:55:38 [pool-7-thread-1] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A] Sending the following host information to the client: dcqos001.usatlas.bnl.gov
22 Apr 2022 14:55:39 [Xrootd-dcqos002-0] [door:Xrootd-dcqos002@xrootd-dcqos002Domain:AAXdQsWeX4A dcqos001_11 DoorTransferFinished 0000DBC6E3CF2C53497ABDF155B4C59A62B5] Transfer 0000DBC6E3CF2C53497ABDF155B4C59A62B5@PoolName=dcqos001_11 PoolAddress=dcqos001_11@dcqos001elevenDomain finished
```

So in conclusion.  Xrootd fails to initiate a redirection using IPV6 IP address. On contrary it works with IP for IPV4.

All the best,
Carlos

**Comment by alrossi** [comment]
Carlos,

(I'm actually on vacation, won't be back until Tuesday, but I took a peek anyway ...)

So what you are saying is the xrootd (xrdcp) client fails if given an ipV6 address, but succeeds if it has an ipV4 address or the canonical name.

I don't yet understand whether this is our problem or not.   What the xrootd door returns is the way dCache resolves the address into a name (or not).   If this is something that should be solved in dCache, it seems to me it is more general than the xroot protocol.

Perhaps another dCache team member could have a look at this before I return  ...

Al

**Comment by DmitryLitvintsev** [comment]
Yes Carlos, 

We have dual stack at Fermilab and we had to add IPV6/host map to `/etc/hosts` on all pool nodes (and door nodes). 

Dmitry

**Comment by DmitryLitvintsev** [comment]
Alternatively you can specify:

```
dcache.java.options.extra=-Djava.net.preferIPv6Addresses=true
```

**Comment by cfgamboa** [comment]
thank you @DmitryLitvintsev

**Comment by alrossi** [comment]
This issue seems to have been resolved for the user.
